docs/investigation/local-hub-v2/dgraph/README.md

```diff
@@ -0,0 +1,112 @@
+
+### Goal
+
+The main goal is to check whether Dgraph can be used as a replacement for current Neo4j implementation to speed up Local Hub extensibility and reduce resource consumption.
+
+Phases:
+1. Check what was changed after a year.
+2. Try to port existing Local Hub in minimal scope:
+	- Query TypeInstances,
+	- and TypInstances mutation with relation.
+3. Use `@custom` or `@lambda` functionality to resolve delegated storage concern.
+	- Instead of official [JavaScript](https://github.com/dgraph-io/dgraph-lambda) implementation, use [dgraph-lambda-go](https://github.com/Schartey/dgraph-lambda-go) server created by community.
+4. If porting was possible, do the benchmarking between current and new solution.
+
+### Non-goal
+
+Even thought that, Public and Local Hubs share the same Neo4j instance, checking if Public Hub can be ported it's **not a goal** of this investigation.
+Additionally, we don't want to use the Dgraph only as of the database and create our own GraphQL server with dedicated resolvers' implementation for each entity.
+
+## Prerequisite
+
+- Install Go
+- Install Docker
+- **(optional)** Install [Insomnia](https://insomnia.rest/download)
+
+## What has changed?
+
+In general, they mostly do a [bug fix releases](https://github.com/dgraph-io/dgraph/blob/12c3ef564cde11ecc3de96ec1516b3148e52d795/CHANGELOG.md).
+
+### Architecture
+
+Dgraph doesn't allow adding custom logic directly in Dgraph server. To change the Dgraph behavior, you can:
+- use `@custom` directive to redirect resolution of a given queries, mutations and fields. Resolution is done by calling the defined HTTP server.
+- use `@lambda` directive to redirect resolution to lambda server. Dgraph provides lambda server only for JavaScript. In general, this is the same concept of executing HTTP calls against a registered Lambda server.
+
+Dgraph is written in Go, but it cannot be directly extended in Go. Custom resolvers are always an HTTP call to your service.
+
+It's due to the fact that Dgraph's GraphQL calls are not translated but directly executed on Dgraph server. In the official docs, you read:
+> Dgraph is the world’s first, and at this time only, service to offer a specification-compliant GraphQL endpoint without the need for an additional translation layer in the tech stack. You will not find a hidden layer of GraphQL resolvers in Dgraph. GraphQL is natively executed within the core of Dgraph itself.
+
+In comparison to current Neo4j implementation. We can do exactly the same, but we need to write delegation logic in JavaScript, as a result we can use any protocol we want to. Dgraph just gives a syntax sugar for extensions, but at the same time restrict it to HTTP only.
+
+#### Give it a try
+
+1. Ensure that Docker is running on your localhost.
+2. Run the Dgraph and lambda server and load GraphQL schema:
+   ```bash
+   docker compose up --build
```

Comment:
I also needed to run
```
docker compose run populate
```

Because of some race condition:

```
dgraph-populate-1  |   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
dgraph-populate-1  |                                  Dload  Upload   Total   Spent    Left  Speed
dgraph-dgraph-1    | I0225 10:28:38.472114      27 middlewares.go:210] GraphQL admin mutation. Name =  updateGQLSchema
100   938  100    87  100   851  42130   402k --:--:-- --:--:-- --:--:--  9{"errors":[{"message":"Unavailable: Server not ready.","extensions":{"code":"Error"}}]}16k
dgraph-populate-1 exited with code 0
```